Zend/zend_timer.c

```diff
@@ -0,0 +1,104 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Kévin Dunglas <kevin@dunglas.dev>                            |
+   +----------------------------------------------------------------------+
+ */
+
+#ifdef ZEND_TIMER
+
+#include <stdio.h>
+#include <signal.h>
+#include <time.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+
+#include "zend.h"
+#include "zend_globals.h"
+
+// Musl Libc defines this macro, glibc does not
+// According to "man 2 timer_create" this field should always be available, but it's not: https://sourceware.org/bugzilla/show_bug.cgi?id=27417
+# ifndef sigev_notify_thread_id
+# define sigev_notify_thread_id _sigev_un._tid
+# endif
+
+ZEND_API void zend_timer_create(void) /* {{{ */
+{
+	struct sigevent sev;
+	sev.sigev_notify = SIGEV_THREAD_ID;
+	sev.sigev_value.sival_ptr = &EG(timer);
+	// The chosen signal must:
+	// 1. not be used internally by libc
+	// 2. be allowed to happen spuriously without consequences
+	// 3. not be commonly used by applications, this excludes SIGALRM, SIGUSR1 and SIGUSR2
+	// 4. not be used by profilers, this excludes SIGPROF
+	// 5. not be used internally by runtimes of programs that can embed PHP, this excludes SIGURG, which is used by Go
+	sev.sigev_signo = SIGIO;
+	sev.sigev_notify_thread_id = (pid_t) syscall(SYS_gettid);
+
+	if (timer_create(CLOCK_THREAD_CPUTIME_ID, &sev, &EG(timer)) != 0) {
```

Comment:
IMO this should be wall-time based, not cpu-time based. Based on this RFC, it's not clear cut on exactly what platforms it already is wall-time based: https://wiki.php.net/rfc/max_execution_wall_time.

The reasoning is pretty simple: it's odd that waiting on a database or sleeping doesn't trigger the timeout when you'd expect.

Reply:
I tried to keep the behavior similar to the existing non-ZTS one, but I agree with you that to me it makes more sense to check the wall-time.

Reply:
We should probably keep the same behavior in ZTS and non-ZTS builds, WDYT? If #6504 gets merged, we'll add support for the new setting too.